projects/ui/src/Utility/AdminUtility/RateLimitEditor.tsx

```diff
@@ -0,0 +1,208 @@
+import { Box, Flex, Input, NumberInput, Text } from "@mantine/core";
+import { FormEvent, useRef, useState } from "react";
+import toast from "react-hot-toast";
+import {
+  CreateUpdateAppMetadataParams,
+  CreateUpdateSubscriptionMetadataParams,
+  useCreateAppMetadataMutation,
+  useCreateSubscriptionMetadataMutation,
+  useUpdateAppMetadataMutation,
+  useUpdateSubscriptionMetadataMutation,
+} from "../../Apis/gg_hooks";
+import { Button } from "../../Components/Common/Button";
+import { useIsAdmin } from "../../Context/AuthContext";
+import { shallowEquals } from "../utility";
+import { SharedMetadataProps } from "./MetadataDisplay";
+
+export const RateLimitEditor = ({
+  item,
+  isEditingRateLimit,
+  customMetadata,
+  rateLimitInfo,
+  onIsEditingRateLimitChange,
+}: SharedMetadataProps & {
+  isEditingRateLimit: boolean;
+  onIsEditingRateLimitChange: (newIsEditinRateLimit: boolean) => void;
+}) => {
+  //
+  //  region State
+  //
+  const initialRateLimitInfo = rateLimitInfo ?? {
+    // This is the default if no rate limit is specified.
+    requestsPerUnit: "0",
+    unit: "UNKNOWN",
+  };
+  let initialRPU = 0;
+  try {
+    initialRPU = Number.parseInt(initialRateLimitInfo.requestsPerUnit ?? "0");
+  } catch {}
+  const formRef = useRef<HTMLFormElement>(null);
+  const [requestsPerUnit, setRequestsPerUnit] = useState(initialRPU);
+  const [unit, setUnit] = useState(initialRateLimitInfo.unit);
+  const requestsPerUnitRef = useRef<HTMLInputElement | null>(null);
+  const isAdmin = useIsAdmin();
+
+  //
+  //  region Saving Data
+  //
+  const { trigger: createAppMetadata } = useCreateAppMetadataMutation();
+  const { trigger: updateAppMetadata } = useUpdateAppMetadataMutation();
+  const { trigger: createSubscriptionMetadata } =
+    useCreateSubscriptionMetadataMutation();
+  const { trigger: updateSubscriptionMetadata } =
+    useUpdateSubscriptionMetadataMutation();
+  const onSave = async (e: FormEvent) => {
+    e.preventDefault();
+    const newRateLimitInfo = {
+      requestsPerUnit: requestsPerUnit.toString(),
+      unit,
+    };
+    const newCustomMetadata = customMetadata ?? {};
+    // Stop here if the object wasn't edited.
+    if (shallowEquals(newRateLimitInfo, initialRateLimitInfo)) {
+      return;
+    }
+    (async () => {
```

Comment:
Small nit that this section has a LOT of repeat that could be done as like `toastObject(verb: string, metadataType:string)` or whatever

Reply:
Agreed - I can do that refactor in a follow up.